### PR TITLE
feat: add `ml/cluster/strided/dkmeansld`

### PR DESCRIPTION
**Pull request by nakul-krishnakumar**
Resolves None.

## Description

> What is the purpose of this pull request?

This pull request:

-   Add `ml/cluster/strided/dkmeansld`.

## Related Issues

> Does this pull request have any related issues?

This pull request has the following related issues:

-   None.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

### AI Assistance

> When authoring the changes proposed in this PR, did you use any kind of AI assistance?

-   [ ] Yes
-   [x] No

If you answered "yes" above, how did you use AI assistance?

-   [ ] Code generation (e.g., when writing an implementation or fixing a bug)
-   [ ] Test/benchmark generation
-   [ ] Documentation (including examples)
-   [ ] Research and understanding

#### Disclosure

> If you answered "yes" to using AI assistance, please provide a short disclosure indicating how you used AI assistance. This helps reviewers determine how much scrutiny to apply when reviewing your contribution. Example disclosures: "This PR was written primarily by Claude Code." or "I consulted ChatGPT to understand the codebase, but the proposed changes were fully authored manually by myself.".

None.

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by nakul-krishnakumar** [comment]
This PR is WIP.

**Comment by nakul-krishnakumar** [comment]
Expected API Design:
image

File: lib/node_modules/@stdlib/ml/cluster/strided/dkmeansld/lib/main.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+* @param {integer} strideX2 - stride of the second dimension.
+* @param {integer} offsetX - starting index.
+* @param {Float64Array} init - strided array containing initial centroid locations.
+* @param {integer} strideInit1 - stride of first dimension.
+* @param {integer} strideInit2 - stride of second dimension.
+* @param {integer} strideInit3 - stride of the third dimension.
+* @param {integer} offsetInit - initial index.
+* @ returns {Result} results object
+*/
+function dkmeansld( M, N, k, replicates, metric, maxIter, tol, X, strideX1, strideX2, offsetX, init, strideInit1, strideInit2, strideInit3, offsetInit ) { // eslint-disable-line max-len
```
Comment (kgryte) [inline lib/node_modules/@stdlib/ml/cluster/strided/dkmeansld/lib/main.js]: I realized after looking at this and the internal array allocation that, similar to https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/stats/strided/dztest, we'll want an `out` parameter for a pre-allocated results object. In the C implementation, we don't want to be dynamically allocating memory.

Comment (kgryte) [reply lib/node_modules/@stdlib/ml/cluster/strided/dkmeansld/lib/main.js]: I am also realizing that we'll want a `samples` (number of samples, M) and a `features` (number of features, N) property on the results object. Having these values will allow consumers to properly read the `init`, `centroids`, and `statistics` linear memory buffers, independent of the original function invocation.

Comment (kgryte) [reply lib/node_modules/@stdlib/ml/cluster/strided/dkmeansld/lib/main.js]: Current thinking is the following signature:

```js
function dkmeansld( M, N, k, replicates, metric, maxIter, tol, x, strideX1, strideX2, offsetX, init, strideInit1, strideInit2, strideInit3, offsetInit, centroids, strideC1, strideC2, offsetC, statistics, strideS1, strideS2, offsetS, labels, strideL1, offsetL, out )
```

where `centroids`, `statistics`, and `labels` are output arrays and `out` is a "results" object with the following fields:

- replicates: number of times the data was clustered with different initial centroids.
- replicate: index of the initial centroids which produced the best result.
- metric: metric name.
- iterations: number of iterations for best results.
- method: algorithm name (e.g., 'lloyd').
- inertia: sum of squared distances to the closest centroid for all samples.
- k: number of clusters.
- samples: number of samples (i.e., M).
- features: number of features (i.e., N).

Similar to `ztest`, several of the parameter values should be copied over to what will be an uninitialized results object, such as M, N, k.

Comment (kgryte) [reply lib/node_modules/@stdlib/ml/cluster/strided/dkmeansld/lib/main.js]: Signature with abbreviated parameter names:

```js
function dkmeansld( M, N, k, rep, metric, maxIter, tol, x, sx1, sx2, ox, init, si1, si2, si3, oi, c, sc1, sc2, oc, stats, ss1, ss2, os, labels, sl, ol, out )
```

Comment (nakul-krishnakumar) [reply lib/node_modules/@stdlib/ml/cluster/strided/dkmeansld/lib/main.js]: > Current thinking is the following signature:
> 
> ```js
> function dkmeansld( M, N, k, replicates, metric, maxIter, tol, x, strideX1, strideX2, offsetX, init, strideInit1, strideInit2, strideInit3, offsetInit, centroids, strideC1, strideC2, offsetC, statistics, strideS1, strideS2, offsetS, labels, strideL1, offsetL, out )
> ```
> 
> where `centroids`, `statistics`, and `labels` are output arrays and `out` is a "results" object with the following fields:
> 
>     * replicates: number of times the data was clustered with different initial centroids.
> 
>     * replicate: index of the initial centroids which produced the best result.
> 
>     * metric: metric name.
> 
>     * iterations: number of iterations for best results.
> 
>     * method: algorithm name (e.g., 'lloyd').
> 
>     * inertia: sum of squared distances to the closest centroid for all samples.
> 
>     * k: number of clusters.
> 
>     * samples: number of samples (i.e., M).
> 
>     * features: number of features (i.e., N).
> 
> 
> Similar to `ztest`, several of the parameter values should be copied over to what will be an uninitialized results object, such as M, N, k.

This function signature did not make sense to me. Could you please clarify if `dkmeansld` is supposed run lloyd algorithm for all `replicates` or only a single iteration.
1. If it is supposed to run a single iteration, then why do we need `replicates` as a function parameter?
2. If it is supposed to run for all `replicates`, then how can we pass the `centroid` matrix as I believe it is supposed to be calculated for each `replicate` separately. [sklearn reference](https://github.com/scikit-learn/scikit-learn/blob/d3898d9d57aeb1e960d266613a2e31b07bca39d7/sklearn/cluster/_kmeans.py#L1505)


In short, my confusion revolves around where this [for loop](https://github.com/scikit-learn/scikit-learn/blob/d3898d9d57aeb1e960d266613a2e31b07bca39d7/sklearn/cluster/_kmeans.py#L1503) iterating over the `replicates` would live, inside `dkmeansld` or outside it?

@kgryte

**Comment by nakul-krishnakumar** [comment]
@kgryte Could you please specify what should go into `statistics` matrix?

File: lib/node_modules/@stdlib/ml/cluster/strided/dkmeansld/lib/main.js (hunk 1 above)
Comment (nakul-krishnakumar) [reply lib/node_modules/@stdlib/ml/cluster/strided/dkmeansld/lib/main.js]: I believe I have figured out my above doubt regarding `replicates` iteration.
The current implementation does the iteration through replicates inside the `dkmeansld` method. Also as far as I understood, `centroid`, `init`, `labels` are all initialized array buffers which are to be modified inside `dkmeansld`. I hope my intuition is correct.
Also please let me know if I should modularize the current implementation into smaller helper functions. Thank you :)
